### ISSUE TITLE
[java] Relax AbstractClassWithoutAnyMethod when class is annotated by @AutoValue

### ISSUE THREAD

**Issue by vanniktech** [pmd/pmd#438]
**Rule Set:**
```xml
<rule ref="rulesets/java/design.xml"/>
```

**Description:**
I have an abstract class that implements an interface. The implementation will be generated by [AutoValue](https://github.com/google/auto/tree/master/value).

**Code Sample demonstrating the issue:**

```java
@AutoValue public abstract class Something implements Some {
}
```

**Running PMD through:** *[Gradle]*

How do you guys feel about relaxing `AbstractClassWithoutAnyMethod` if you have an abstract class that implements an interface? If you feel like it's to much of a relaxation we could also apply the constraint of having an `@AutoValue` annotation on the type.

**Comment by jsotuyod** [comment]
@vanniktech thanks for the suggestion.

Not flagging any class that implements an interface seems a little too relaxed. Given no annotations, that leaves you in one of 2 possible scenarios:
1. the interface is a marker interface with no methods, in which case having the abstract class adds no value; and actually hampers developers due to the *"extends one, implement many"* restrictions of the language. Going for the interface on this scenario is **always** better. Flagging these is good for developers.
2. the interface does define methods, but the class doesn't define any. This leaves users with the need to always extend it or write something along the lines of:
```
new Something() {
  // Maybe some @Override methods, maybe not if the interface defines default implementations.
}
```

Replacing `Something` for `Some` works exactly the same way, so once again, the class provides no value over the interface. Flagging these is good for developers.


On the other hand, whitelisting `@AutoValue`, as we have whitelisted Lombok annotations on several rules seems far more reasonable, since we know semantically how that behaves and how it affects the outcome.

**Comment by vanniktech** [comment]
> Replacing Something for Some works exactly the same way, so once again, the class provides no value over the interface. Flagging these is good for developers.

Not quite. In my case `Some` is a generated interface. In order to let AutoValue pick it up I need to create my own class.

Whitelisting `@AutoValue` regarding this rule, sounds good to me.

**Comment by jsotuyod** [comment]
As I said,
> **Given no annotations**, that leaves you in one of 2 possible scenarios:

But I'm glad we seem to have agreed that whitelisting the annotation is the way to go. Feel free to submit a PR if this is something urgent to you, I don't feel we can commit it to 5.8.0 otherwise.
